fmt: panic on type error

An edge case for sure, but `opa fmt` panics on cerrtain invalid constructs, like this example.

```rego
package play

test_and_false_false {
	and([true, false]) == false
}
```

```
❯ opa fmt .
panic: runtime error: index out of range [2] with length 2

goroutine 1 [running]:
github.com/open-policy-agent/opa/format.(*writer).writeCall(0x14000123a40, 0x0, {0x140002f2cf0?, 0x2, 0x2}, {0x0, 0x0, 0x0})
	/Users/anderseknert/git/opa/opa/format/format.go:781 +0x248
github.com/open-policy-agent/opa/format.(*writer).writeTermParens(0x14000123a40, 0x0?, 0x1400026bd10, {0x0?, 0x140003bf248?, 0x102dbb7ac?})
	/Users/anderseknert/git/opa/opa/format/format.go:711 +0x2b8
github.com/open-policy-agent/opa/format.(*writer).writeTerm(...)
	/Users/anderseknert/git/opa/opa/format/format.go:676
github.com/open-policy-agent/opa/format.(*writer).writeFunctionCall(0x14000123a40, 0x0?, {0x0, 0x0, 0x0})
	/Users/anderseknert/git/opa/opa/format/format.go:638 +0x2bc
github.com/open-policy-agent/opa/format.(*writer).writeExpr(0x14000123a40, 0x140001aef40, {0x0?, 0x0?, 0x140001aec00?})
	/Users/anderseknert/git/opa/opa/format/format.go:541 +0x108
github.com/open-policy-agent/opa/format.(*writer).writeBody(0x14000123a40?, {0x140005d84b8, 0x1, 0x0?}, {0x0?, 0x0?, 0x10403a520?})
	/Users/anderseknert/git/opa/opa/format/format.go:519 +0x80
github.com/open-policy-agent/opa/format.(*writer).writeRule(0x14000123a40, 0x140001aef00, 0x0?, {0x0?, 0x0?, 0x0?}, {0x0, 0x0, 0x0})
	/Users/anderseknert/git/opa/opa/format/format.go:358 +0x3bc
github.com/open-policy-agent/opa/format.(*writer).writeRules(0x10378dca0?, {0x140005d84f0, 0x1, 0x0?}, {0x0?, 0x0?, 0x0?}, {0x0?, 0x10460c108?, 0x12bc4db98?})
	/Users/anderseknert/git/opa/opa/format/format.go:305 +0x84
github.com/open-policy-agent/opa/format.(*writer).writeModule(0x14000123a40, 0x1400049b180?, {0x0?, 0xec?, 0x27?})
	/Users/anderseknert/git/opa/opa/format/format.go:272 +0x3e8
github.com/open-policy-agent/opa/format.AstWithOpts({0x103737020?, 0x1400049b100?}, {0x40?})
	/Users/anderseknert/git/opa/opa/format/format.go:152 +0x590
github.com/open-policy-agent/opa/format.Ast(...)
	/Users/anderseknert/git/opa/opa/format/format.go:61
github.com/open-policy-agent/opa/format.Source({0x140001e9f30, 0xb}, {0x14000228800?, 0x1027e4714?, 0x140003bf898?})
	/Users/anderseknert/git/opa/opa/format/format.go:40 +0xc8
github.com/open-policy-agent/opa/cmd.formatFile(0x104075058, {0x103822260, 0x14000120008}, {0x140001e9f30, 0xb}, {0x10382e6c8, 0x1400011f450}, {0x0?, 0x0?})
	/Users/anderseknert/git/opa/opa/cmd/fmt.go:111 +0x1d4
github.com/open-policy-agent/opa/cmd.opaFmt.func1({0x140001e9f30?, 0x0?}, {0x10382e6c8?, 0x1400011f450?}, {0x0?, 0x0?})
	/Users/anderseknert/git/opa/opa/cmd/fmt.go:76 +0x68
path/filepath.walk({0x140001e9f30, 0xb}, {0x10382e6c8, 0x1400011f450}, 0x103819748)
	/opt/homebrew/Cellar/go/1.19.4/libexec/src/path/filepath/path.go:433 +0xd0
path/filepath.walk({0x16d6835f8, 0x1}, {0x10382e6c8, 0x1400011f380}, 0x103819748)
	/opt/homebrew/Cellar/go/1.19.4/libexec/src/path/filepath/path.go:457 +0x1fc
path/filepath.Walk({0x16d6835f8, 0x1}, 0x103819748)
	/opt/homebrew/Cellar/go/1.19.4/libexec/src/path/filepath/path.go:520 +0x6c
github.com/open-policy-agent/opa/cmd.opaFmt({0x140002f2bc0?, 0x1, 0x0?})
	/Users/anderseknert/git/opa/opa/cmd/fmt.go:75 +0xe4
github.com/open-policy-agent/opa/cmd.glob..func1(0x104020d60?, {0x140002f2bc0?, 0x1?, 0x1?})
	/Users/anderseknert/git/opa/opa/cmd/fmt.go:53 +0x2c
github.com/spf13/cobra.(*Command).execute(0x104020d60, {0x140002f2b80, 0x1, 0x1})
	/Users/anderseknert/git/opa/opa/vendor/github.com/spf13/cobra/command.go:920 +0x5b0
github.com/spf13/cobra.(*Command).ExecuteC(0x104020a80)
	/Users/anderseknert/git/opa/opa/vendor/github.com/spf13/cobra/command.go:1044 +0x35c
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/anderseknert/git/opa/opa/vendor/github.com/spf13/cobra/command.go:968
main.main()
	/Users/anderseknert/git/opa/opa/main.go:14 +0x28
```

Originally reported by @nevumx.

--- a/cmd/fmt_test.go
+++ b/cmd/fmt_test.go
@@ -2,12 +2,15 @@ package cmd
 
 import (
 	"bytes"
+	"fmt"
 	"io"
 	"os"
 	"path/filepath"
 	"strings"
 	"testing"
 
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/format"
 	"github.com/open-policy-agent/opa/util/test"
 )
 
@@ -28,6 +31,27 @@ const unformatted = `
         }
 
 
+`
+
+const singleWrongArity = `package test
+
+p {
+	a := 1
+	b := 2
+	plus(a, b, c) == 3
+}
+`
+
+const MultipleWrongArity = `package test
+
+p {
+	x:=5
+	y:=7
+	z:=6
+	plus([x, y]) == 3
+	and(true, false, false) == false
+	plus(a, x, y, z)
+}
 `
 
 func TestFmtFormatFile(t *testing.T) {
@@ -266,6 +290,92 @@ func TestFmtFailFileChangesDiff(t *testing.T) {
 		actual := strings.TrimSpace(stdout.String())
 		if len(actual) == 0 {
 			t.Fatalf("Expected output, got:\n%v\n\n", actual)
+		}
+	})
+}
+
+func TestFmtSingleWrongArityError(t *testing.T) {
+	params := fmtCommandParams{}
+	var stdout bytes.Buffer
+
+	files := map[string]string{
+		"policy.rego": singleWrongArity,
+	}
+
+	test.WithTempFS(files, func(path string) {
+		policyFile := filepath.Join(path, "policy.rego")
+		info, err := os.Stat(policyFile)
+		err = formatFile(&params, &stdout, policyFile, info, err)
+		if err == nil {
+			t.Fatalf("Expected error but did not receive one")
+		}
+
+		loc := ast.Location{File: policyFile, Row: 6}
+		errExp := ast.NewError(ast.TypeErr, &loc, "%s: %s", "plus", "arity mismatch")
+		errExp.Details = &format.ArityFormatErrDetail{
+			Have: []string{"var", "var", "var"},
+			Want: []string{"number", "number"},
+		}
+		expectedErrs := ast.Errors(make([]*ast.Error, 1))
+		expectedErrs[0] = errExp
+		expectedSingleWrongArityErr := newError("failed to parse Rego source file: %v", fmt.Errorf("%s: %v", policyFile, expectedErrs))
+
+		if err != expectedSingleWrongArityErr {
+			t.Fatalf("Expected:%s\n\nGot:%s\n\n", expectedSingleWrongArityErr, err)
+		}
+	})
+}
+
+func TestFmtMultipleWrongArityError(t *testing.T) {
+	params := fmtCommandParams{}
+	var stdout bytes.Buffer
+
+	files := map[string]string{
+		"policy.rego": MultipleWrongArity,
+	}
+
+	test.WithTempFS(files, func(path string) {
+		policyFile := filepath.Join(path, "policy.rego")
+		info, err := os.Stat(policyFile)
+		err = formatFile(&params, &stdout, policyFile, info, err)
+		if err == nil {
+			t.Fatalf("Expected error but did not receive one")
+		}
+
+		locations := []ast.Location{
+			{File: policyFile, Row: 7},
+			{File: policyFile, Row: 8},
+			{File: policyFile, Row: 9},
+		}
+		haveStrings := [][]string{
+			{"array"},
+			{"boolean", "boolean", "boolean"},
+			{"var", "var", "var", "var"},
+		}
+		wantStrings := [][]string{
+			{"number", "number"},
+			{"set[any]", "set[any]"},
+			{"number", "number"},
+		}
+		operators := []string{
+			"plus",
+			"and",
+			"plus",
+		}
+		expectedErrs := ast.Errors(make([]*ast.Error, 3))
+		for i := 0; i < 3; i++ {
+			loc := locations[i]
+			errExp := ast.NewError(ast.TypeErr, &loc, "%s: %s", operators[i], "arity mismatch")
+			errExp.Details = &format.ArityFormatErrDetail{
+				Have: haveStrings[i],
+				Want: wantStrings[i],
+			}
+			expectedErrs[i] = errExp
+		}
+		expectedMultipleWrongArityErr := newError("failed to parse Rego source file: %v", fmt.Errorf("%s: %v", policyFile, expectedErrs))
+
+		if err != expectedMultipleWrongArityErr {
+			t.Fatalf("Expected:%s\n\nGot:%s\n\n", expectedMultipleWrongArityErr, err)
 		}
 	})
 }

--- a/format/format.go
+++ b/format/format.go
@@ -10,9 +10,11 @@ import (
 	"fmt"
 	"regexp"
 	"sort"
+	"strings"
 
 	"github.com/open-policy-agent/opa/ast"
 	"github.com/open-policy-agent/opa/internal/future"
+	"github.com/open-policy-agent/opa/types"
 )
 
 // Opts lets you control the code formatting via `AstWithOpts()`.
@@ -37,6 +39,7 @@ func Source(filename string, src []byte) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
+
 	formatted, err := Ast(module)
 	if err != nil {
 		return nil, fmt.Errorf("%s: %v", filename, err)
@@ -142,6 +145,7 @@ func AstWithOpts(x interface{}, opts Opts) ([]byte, error) {
 
 	w := &writer{
 		indent: "\t",
+		errs:   make([]*ast.Error, 0),
 	}
 
 	switch x := x.(type) {
@@ -178,6 +182,9 @@ func AstWithOpts(x interface{}, opts Opts) ([]byte, error) {
 		return nil, fmt.Errorf("not an ast element: %v", x)
 	}
 
+	if len(w.errs) > 0 {
+		return nil, w.errs
+	}
 	return squashTrailingNewlines(w.buf.Bytes()), nil
 }
 
@@ -228,6 +235,7 @@ type writer struct {
 	inline    bool
 	beforeEnd *ast.Comment
 	delay     bool
+	errs      ast.Errors
 }
 
 func (w *writer) writeModule(module *ast.Module, o fmtOpts) {
@@ -587,7 +595,7 @@ func (w *writer) writeSomeDecl(decl *ast.SomeDecl, comments []*ast.Comment) []*a
 				w.write(",")
 			}
 		case ast.Call:
-			comments = w.writeInOperator(false, val[1:], comments)
+			comments = w.writeInOperator(false, val[1:], comments, decl.Location, ast.BuiltinMap[val[0].String()].Decl)
 		}
 	}
 
@@ -622,7 +630,7 @@ func (w *writer) writeFunctionCall(expr *ast.Expr, comments []*ast.Comment) []*a
 
 	switch operator {
 	case ast.Member.Name, ast.MemberWithKey.Name:
-		return w.writeInOperator(false, terms[1:], comments)
+		return w.writeInOperator(false, terms[1:], comments, terms[0].Location, ast.BuiltinMap[terms[0].String()].Decl)
 	}
 
 	bi, ok := ast.BuiltinMap[operator]
@@ -647,6 +655,9 @@ func (w *writer) writeFunctionCall(expr *ast.Expr, comments []*ast.Comment) []*a
 		comments = w.writeTerm(terms[2], comments)
 		return comments
 	}
+	// NOTE(Trolloldem): in this point we are operating with a built-in function with the
+	// wrong arity even when the assignment notation is used
+	w.errs = append(w.errs, ArityFormatMismatchError(terms[1:], terms[0].String(), terms[0].Location, bi.Decl))
 	return w.writeFunctionCallPlain(terms, comments)
 }
 
@@ -708,7 +719,7 @@ func (w *writer) writeTermParens(parens bool, term *ast.Term, comments []*ast.Co
 	case ast.Var:
 		w.write(w.formatVar(x))
 	case ast.Call:
-		comments = w.writeCall(parens, x, comments)
+		comments = w.writeCall(parens, x, term.Location, comments)
 	case fmt.Stringer:
 		w.write(x.String())
 	}
@@ -760,7 +771,7 @@ func (w *writer) formatVar(v ast.Var) string {
 	return v.String()
 }
 
-func (w *writer) writeCall(parens bool, x ast.Call, comments []*ast.Comment) []*ast.Comment {
+func (w *writer) writeCall(parens bool, x ast.Call, loc *ast.Location, comments []*ast.Comment) []*ast.Comment {
 	bi, ok := ast.BuiltinMap[x[0].String()]
 	if !ok || bi.Infix == "" {
 		return w.writeFunctionCallPlain(x, comments)
@@ -769,13 +780,22 @@ func (w *writer) writeCall(parens bool, x ast.Call, comments []*ast.Comment) []*
 	if bi.Infix == "in" {
 		// NOTE(sr): `in` requires special handling, mirroring what happens in the parser,
 		// since there can be one or two lhs arguments.
-		return w.writeInOperator(true, x[1:], comments)
+		return w.writeInOperator(true, x[1:], comments, loc, bi.Decl)
 	}
 
 	// TODO(tsandall): improve to consider precedence?
 	if parens {
 		w.write("(")
 	}
+
+	// NOTE(Trolloldem): writeCall is only invoked when the function call is a term
+	// of another function. The only valid arity is the one of the
+	// built-in function
+	if len(bi.Decl.Args()) != len(x)-1 {
+		w.errs = append(w.errs, ArityFormatMismatchError(x[1:], x[0].String(), loc, bi.Decl))
+		return comments
+	}
+
 	comments = w.writeTermParens(true, x[1], comments)
 	w.write(" " + bi.Infix + " ")
 	comments = w.writeTermParens(true, x[2], comments)
@@ -786,7 +806,16 @@ func (w *writer) writeCall(parens bool, x ast.Call, comments []*ast.Comment) []*
 	return comments
 }
 
-func (w *writer) writeInOperator(parens bool, operands []*ast.Term, comments []*ast.Comment) []*ast.Comment {
+func (w *writer) writeInOperator(parens bool, operands []*ast.Term, comments []*ast.Comment, loc *ast.Location, f *types.Function) []*ast.Comment {
+	if len(operands) != len(f.Args()) {
+		// The number of operands does not math the arity of the `in` operator
+		operator := ast.Member.Name
+		if len(f.Args()) == 3 {
+			operator = ast.MemberWithKey.Name
+		}
+		w.errs = append(w.errs, ArityFormatMismatchError(operands, operator, loc, f))
+		return comments
+	}
 	kw := "in"
 	switch len(operands) {
 	case 2:
@@ -1355,4 +1384,37 @@ func ensureFutureKeywordImport(imps []*ast.Import, kw string) []*ast.Import {
 	}
 	imp.Location = defaultLocation(imp)
 	return append(imps, imp)
+}
+
+// ArgErrDetail but for `fmt` checks since compiler has not run yet.
+type ArityFormatErrDetail struct {
+	Have []string `json:"have"`
+	Want []string `json:"want"`
+}
+
+// arityMismatchError but for `fmt` checks since the compiler has not run yet.
+func ArityFormatMismatchError(operands []*ast.Term, operator string, loc *ast.Location, f *types.Function) *ast.Error {
+	want := make([]string, len(f.Args()))
+	for i := range f.Args() {
+		want[i] = types.Sprint(f.Args()[i])
+	}
+
+	have := make([]string, len(operands))
+	for i := 0; i < len(operands); i++ {
+		have[i] = ast.TypeName(operands[i].Value)
+	}
+	err := ast.NewError(ast.TypeErr, loc, "%s: %s", operator, "arity mismatch")
+	err.Details = &ArityFormatErrDetail{
+		Have: have,
+		Want: want,
+	}
+	return err
+}
+
+// Lines returns the string representation of the detail.
+func (d *ArityFormatErrDetail) Lines() []string {
+	return []string{
+		"have: " + "(" + strings.Join(d.Have, ",") + ")",
+		"want: " + "(" + strings.Join(d.Want, ",") + ")",
+	}
 }

